### PR TITLE
[SPARK-29674][CORE] Update dropwizard metrics to 4.1.x for JDK 9+

### PR DESCRIPTION
**Pull request by srowen**
### What changes were proposed in this pull request?

Update the version of dropwizard metrics that Spark uses for metrics to 4.1.x, from 3.2.x. 

### Why are the changes needed?

This helps JDK 9+ support, per for example https://github.com/dropwizard/metrics/pull/1236

### Does this PR introduce any user-facing change?

No, although downstream users with custom metrics may be affected.

### How was this patch tested?

Existing tests.

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -292,6 +292,16 @@
     <dependency>
       <groupId>io.dropwizard.metrics</groupId>
       <artifactId>metrics-graphite</artifactId>
+      <exclusions>
+        <exclusion>
+          <groupId>com.rabbitmq</groupId>
+          <artifactId>amqp-client</artifactId>
```
Comment (srowen) [inline core/pom.xml]: This would otherwise be a new dependency on RabbitMQ, but, my hunch is that it is not necessary for Spark.

**Comment by dongjoon-hyun** [comment]
Wow. Nice!

**Comment by dongjoon-hyun** [comment]
BTW, what happens to `ganglia`? Do we still have `Ganglia` support after this PR?

**Comment by srowen** [comment]
I don't know of a reason it would change Ganglia support. `metrics-ganglia` is still included, updated, and built here.

**Comment by srowen** [comment]
Oh, wait no I see the issue. Ganglia support is in a different module that I didn't enable. Indeed it isn't present in 4.x: https://search.maven.org/search?q=g:io.dropwizard.metrics%20AND%20a:metrics-ganglia&core=gav

Hm, OK, that's a problem, as we have historically maintained some Ganglia support.
But there are actually no tests for this integration, so we wouldn't have noticed any issues with it anyway.

Maybe I need to take this to dev@ ... if the support doesn't necessarily work on JDK 11 (unproven) then do we need to consider removing it?

**Comment by dongjoon-hyun** [comment]
Since we dropped `kafka-0.8` at 2.4.x (Scala-2.12), can can disable `Ganglia` support in JDK11 first? It's because the main distribution of Apache Spark 3.0.0 is still JDK8.

**Comment by dongjoon-hyun** [comment]
Oh.. It's difficult to keep separately for JDK8/11.

**Comment by LucaCanali** [comment]
Adding a link to a similar discussion in #26212
Probably getting rid of the Ganglia reporter is a good solution. Just a thought is that if some people still need it in Spark 3.x, maybe the Ganglia reporter can be supported by Apache Spark? It does not seem like a lot of code. There is the annoying issue of LGPL licensing of course though, that's why Dropwizard metrics library dropped it, I understand.

**Comment by srowen** [comment]
Yep, I am next going to try that, just inlining the simple single file metrics-ganglia contained. Dropwizard itself is ALv2. Ganglia is not, but, Spark doesn't publish this module directly, so that doesn't change.
Honestly most of the work is updating the license files!

**Comment by srowen** [comment]
Yep, I am next going to try that, just inlining the simple single file metrics-ganglia contained. Dropwizard itself is ALv2. Ganglia is not, but, Spark doesn't publish this module directly, so that doesn't change.
Honestly most of the work is updating the license files!
If this file gets hard to maintain, we can consider removal. If it's pretty easy to keep it for now, we can just keep it.
Honestly the only problem here is lack of tests.

File: external/spark-ganglia-lgpl/pom.xml [hunk 2]
```diff
@@ -39,10 +39,10 @@
       <artifactId>spark-core_${scala.binary.version}</artifactId>
       <version>${project.version}</version>
     </dependency>
-
     <dependency>
-      <groupId>io.dropwizard.metrics</groupId>
-      <artifactId>metrics-ganglia</artifactId>
+      <groupId>info.ganglia.gmetric4j</groupId>
```
Comment (srowen) [inline external/spark-ganglia-lgpl/pom.xml]: This was previously a transitive dependency anyway, and should have been directly referenced.

File: NOTICE-binary [hunk 3]
```diff
@@ -1515,3 +1515,16 @@ Copyright 2014-2017 The Apache Software Foundation
 
 This product includes software developed at
 The Apache Software Foundation (http://www.apache.org/).
+
+
+Metrics
```
Comment (srowen) [inline NOTICE-binary]: I think this should have already been there, but now would also need to be in `NOTICE`

**Comment by dongjoon-hyun** [comment]
IIUC, we are publishing our `Ganglia` module to Maven directly from Apache Spark `1.2.0` to `2.4.4` technically.
> Spark doesn't publish this module directly, so that doesn't change.
- https://repo1.maven.org/maven2/org/apache/spark/spark-ganglia-lgpl_2.11/

**Comment by srowen** [comment]
Ah sorry, you are correct. What I really mean is that it isn't bundled in the binary release, because that would entail distributing the ganglia dependency. At least this was true last time I checked!

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Review by dongjoon-hyun — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
Thank you so much, @srowen . Merged to master.
Also, thank you, @LucaCanali .

**Comment by srowen** [comment]
@LucaCanali @jerryshao maybe I can get your advice on something related here.

I was investigating back-porting this to 2.4.x. I find that the tests run out of memory with this change. A heap dump suggests that a lot of memory in consumed by `MetricRegistry` objects, with transitive references to all sorts of big objects. (Let's put aside why those references exist for now.)

As part of my hacking, I found that manually clearing all the `Metric`s in the registry when Spark's `MetricSystem` stops seems to resolve it. As in, in `MetricSystem`:

```
def stop(): Unit = {
    ...
      sinks.foreach(_.stop)
      registry.removeMatching(new MetricFilter {
        override def matches(name: String, metric: Metric): Boolean = true
      })
    ...
  }
```

Is this crazy? like, is clearing out all of the metrics going to break something? or is this a sensible cleanup step?

I'm still investigating, but wanted to reach out to anyone who may be familiar with this code.

**Comment by LucaCanali** [comment]
What you write sounds reasonable to me. As an additional comment I'd say that while stopping the metrics system makes sense for a testing suite, it is not something that happens often in real usage: for most cases I can think of, metrics are started/registered at the component instance startup and they stay up for all its life.

**Comment by srowen** [comment]
Thanks @LucaCanali . Yeah, that's my worry. This means that after the Spark cluster is stopped, the metrics all disappear. Is that expected, or a problem? I am not sure how metrics are used. Of course at that point the components' lifecycle has ended anyway. Would external services rely on 'pulling' metrics after it's done but before the JVM stops, or do metrics implementations typically 'push' as they go? sorry for the dumb question.

**Comment by LucaCanali** [comment]
Metrics are written to sinks at regular time intervals (configurable, let’s say 10 seconds). I mostly use the Graphite sink to write to an InfluxDB instance. Metrics are persisted there and consumed from InfluxDB by a Grafana dashboard.
The metrics system also has the MetricsServlet and more recently PrometheusServlet sink, which publish metrics to the Spark UI, + also I believe the JmxSink is quite popular. In these cases the idea is for the metrics to be consumed by the final user by pulling. It would be the end user/consumer's job to keep a history of the gathered metrics in these cases.

**Comment by srowen** [comment]
OK it's sounding reasonable to explicitly remove the registry then, as it's going away pretty immediately anyway, and the sinks are shutdown. I will propose it.

That helps the tests, which do start/stop contexts, and I think we otherwise somehow 'leak' all the previous metrics registries in this case. However it's kind of specific to the tests.

The other question remains: why is this an issue only in 2.4 (as far as I can tell), and only with metrics 4.x? Since you've looked at a lot of metrics related changes @LucaCanali , do you recall anything we fixed in 3.0 that might help avoid using a lot of memory in the metrics system? I don't, and didn't see anything in skimming JIRAs.

I am still not clear whether the big closures captured by gauges that Spark registers are necessary or an accident that will otherwise cause long-running apps to 'leak' somehow. I don't think so; what it retains a reference to are all the current Spark machinery like BlockManager. Not a problem as long as there is only one live BlockManager anyway. Still, need to keep an eye on it in 3.0.

**Comment by LucaCanali** [comment]
Most of the PRs for the metrics system targeting 3.0 so far have been about adding new metrics.
It's good to know that the memory usage issue does not seem to affect 3.0 though.
BTW we could work on adding a few configuration parameters to allow turning on/off some of the more "chatty metrics"? For example: #26320

**Pull request closed without merging**